### PR TITLE
Run code blocks of .ipynb files as if they were inside a .py file

### PR DESCRIPTION
**Pull request by mederka**
- Let users add .ipynb files to models in schema.yml
- Process .ipynb files to extract Python code
- Run Python code from .ipynb files the same way as code from .py files
- Make .ipynb scripts work with `fal flow` command

Addresses #288 

Fixes FEA-14

### CODE REVIEW

File: src/fal/node_graph.py [hunk 1]
```diff
@@ -148,13 +148,13 @@ def get_node(self, id: str) -> FalFlowNode | None:
...
+        return _is_script(node_name)
 
     def _is_critical_node(self, node):
         successors = list(self.graph.successors(node))
         if node in successors:
             successors.remove(node)
-        has_script_pred = lambda node_name: node_name.endswith(".py")
+        has_script_pred = _is_script
         has_model_pred = lambda node_name: node_name.split(".")[0] == "model"
         if any(has_script_pred(i) for i in successors) and any(
```
Comment (chamini2) [inline src/fal/node_graph.py:159]: you can just change `has_script_pred` for `_is_script` and remove `has_script_pred` var.

Comment (chamini2) [inline src/fal/node_graph.py:159]: you can just change `has_script_pred` for `_is_script` and remove `has_script_pred` var.

**Review by chamini2 — COMMENTED** [review]
I am _worried_ that we are simply just running the code, and notebooks I think will usually have
```py
from fal import FalDbt
faldbt = FalDbt(profiles_dir=..., project_dir=...)
faldbt.<stuff>
```

What I mean is, that these scripts are ignoring the injected context (ref, source, etc) and just running not-really-related code after their models.

*****

A problematic example is if you do:
```
$ fal flow run --target other_profile
```
This will run all the dbt stuff against `other_profile`, but the script that just loads a fal project will not get that information, so it will actually connect to the default profile to read and do changes there.

**Comment by turbo1912** [comment]
> I am _worried_ that we are simply just running the code, and notebooks I think will usually have
> 
> ```python
> from fal import FalDbt
> faldbt = FalDbt(profiles_dir=..., project_dir=...)
> faldbt.<stuff>
> ```
> 
> What I mean is, that these scripts are ignoring the injected context (ref, source, etc) and just running not-really-related code after their models.
> 
> A problematic example is if you do:
> 
> ```
> $ fal flow run --target other_profile
> ```
> 
> This will run all the dbt stuff against `other_profile`, but the script that just loads a fal project will not get that information, so it will actually connect to the default profile to read and do changes there.

How can we avoid this though? For example does this not happen in their notebooks as well? Seems like this is something people should be aware of, happy to discuss it more.. If someone chooses to use fal as a library some of the nice ergonomics go away..

File: src/fal/fal_script.py [hunk 2]
```diff
@@ -157,3 +160,19 @@ def _process_tests(tests: List[Any]):
             tests,
         )
     )
+
+
+def _process_ipynb(filepath: str) -> str:
+    def strip_magic(source: str) -> str:
```
[the remark this answers is not in the document]
Comment (mederka) [reply src/fal/fal_script.py:166]: We're returning a string here

**Comment by mederka** [comment]
> I am _worried_ that we are simply just running the code, and notebooks I think will usually have
> 
> ```python
> from fal import FalDbt
> faldbt = FalDbt(profiles_dir=..., project_dir=...)
> faldbt.<stuff>
> ```
> 
> What I mean is, that these scripts are ignoring the injected context (ref, source, etc) and just running not-really-related code after their models.
> 
> A problematic example is if you do:
> 
> ```
> $ fal flow run --target other_profile
> ```
> 
> This will run all the dbt stuff against `other_profile`, but the script that just loads a fal project will not get that information, so it will actually connect to the default profile to read and do changes there.

Thank you for reviewing this PR, @chamini2! 🙏 Totally understand your worries. Actually ran into the same issue when writing integration tests for this feature. After a bit of thinking though, I came to the same conclusions as @turbo1912.

In the case of notebooks, the Python cells should run the same way as they do in a notebook. If the user wants the flexibility of a Python script, they should convert the notebook to a Python script themselves and handle the magic functions.

A potentially interesting feature would be to offer magic functions and variables in a notebook, such that they get the same flexibility but in notebook runtime. We would have to parse the dbt project outside of fal runtime. That would be an adventure 😄 .

File: src/fal/fal_script.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chamini2) [reply src/fal/fal_script.py:167]:
```py
    for cell in raw_script["cells"]:
        if cell["cell_type"] == "code":
            source = strip_magic(cell["source"])
            script_list.append("".join(source))
```

`cell["source]"` is a `List[str]`

and you are doing `... for item in source ...` isn't `source` a list?

**Comment by mederka** [comment]
> Maybe we could set a special variable in the context to indicate "you are in the runtime" when in the fal runtime. And then somehow check for the var while constructing a `FalDbt` object. I don't know if it would be fragile, but just an idea.

In any case, calls for a design doc!

**Review by chamini2 — APPROVED** [review]
amazing stuff

**Review by turbo1912 — APPROVED** [review]

**Pull request merged**